source/extensions/filters/http/oauth2/oauth_response.proto

```diff
@@ -6,5 +6,11 @@ import "google/protobuf/wrappers.proto";
 
 message OAuthResponse {
   google.protobuf.StringValue access_token = 1;
+
   google.protobuf.UInt64Value expires_in = 2;
+
+  // Subject of the token (user)
+  google.protobuf.StringValue sub = 3;
+
+  repeated google.protobuf.StringValue groups = 4;
```

Comment:
This assumes the response to include the field `groups` but it seems `groups` (and `sub`) is not in the standard for a successful OAuth response: https://tools.ietf.org/html/rfc6749#section-5.1. (`access_token` and `expires_in` is in the standard)

It looks like keycloak will include the `groups` field but I'm wondering should envoy has a more general solution that works for other provider that doesn't use the `groups` field?

Reply:
For that conventional OAuth standard, yes, it is expected that the grantee perform another call to the authentication endpoint, using the token as a parameter, to obtain further `groups` information. However I do believe the future of user authentication should instead leverage the more modernized claim transfer techniques such as JWT, which contains `groups` and `sub`. I think it's acceptable to extract out the groups directly from the response if available because that is the preferred token format.

If there is significant adoption of the older https://tools.ietf.org/html/rfc6749#section-5.1 I can create a follow-up that can inform the OAuth filter how to exactly obtain `groups` info.

Reply:
@williamsfu99 Is there any specification that explains the `groups` field? It looks a non-standard field specific to your OAuth 2.0 authorization server. I can't find any mention of the `groups` field in [RFC 6749 Section 5.1](https://tools.ietf.org/html/rfc6749#section-5.1) nor [OpenID Connect Core 1.0 Section 3.1.3.3](https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse).

There's a good chance this feature is incompatible with most major OAuth 2.0 providers. GitHub, for example, runs their identity layer (username, group membership, etc) separately from OAuth 2.0.
OpenID Connect is arguably the most widespread OAuth 2.0 identity layer, and it's only used by a handful of providers: https://en.wikipedia.org/wiki/List_of_OAuth_providers

cc @rgs1 @yangminzhu

Reply:
> However I do believe the future of user authentication should instead leverage the more modernized claim transfer techniques such as JWT, which contains groups and sub. I think it's acceptable to extract out the groups directly from the response if available because that is the preferred token format.

The `sub` claim is standard per RFC 7519 (JWT), but `groups` isn't.

FWIW OpenID Connect (OIDC) uses JWT for transferring claims, but not in the root level of the `access_token` response. It places the encoded JWT into the `id_token` field.

What do you think of an opt-in to export the JWT contained in the OIDC `id_token` through metadata or request headers? Any further filters (like Lua, ext_authz) could then apply arbitrary access control based on JWT claims.

Reply:
Thanks for all the clarification. I'm interested in this as I was just experimenting enforcing access control based on the group. I'm testing with Google OAuth and the group information is not in the `access_token` and need to be retried by calling a separate API.

> What do you think of an opt-in to export the JWT contained in the OIDC id_token through metadata or request headers? Any further filters (like Lua, ext_authz) could then apply arbitrary access control based on JWT claims.

This should be very useful and gives more flexibility about the access control. I think even the `access_response` could be stored in the metadata so that it could be consumed by later filters. for example, the RBAC filter supports the very powerful metadata matcher for such use cases. But this might be out the scope of this PR and the user experience will also be a little bit more complicated. 

For this PR, maybe use the [metadata matcher](https://github.com/envoyproxy/envoy/blob/master/api/envoy/type/matcher/v3/metadata.proto#L78) instead of   `authorized_groups` field? This implies the JWT will also be stored to the dynamic metadata, and gives us more flexibility so that we can use it to match against different fields. 
That's just an idea, I'm fine with either way and will let the Envoy maintainer to decide on this. Thanks.

Reply:
Sorry for the delays.

Due to arguments about this filter's inconsistence with authn/authz separation and its incompatibility with other authn techniques, I'll abandon this PR. 

Thanks for the feedback around specialized logic that tailors better to generic JWT structures and maintaining their contents in metadata. However I notice the [official JWT filter](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/http/jwt_authn/v2alpha/config.proto) has an option to write payload data into metadata already. I think that filter can be updated to support specific behaviors shared in this thread.

I think the only thing that makes sense as an improvement to the current filter is having an option to fetch groups/user information from a separately configured API (e.g. useful for Google OAuth) that is then stored in metadata/passed along as headers. Fetching authenticated identification info is a reasonable expectation for an authn filter and does not conflate authn & authz. 

cc @rgs1 and @juanvasquezreyes who originally filed the request.